### PR TITLE
refactor(methods): promote use of traits, double down on the macro complexity

### PR DESCRIPTION
**Pull request by miraclx**
This PR;

- Implements new traits `RpcHandlerRequest` and `RpcHandlerError` on all traits across the board for each method. Allowing for smarter static type checking and request/error transformers.
- Doubles down on the previously heavy usage of macros.
  - As a consequence of this, we're also dropping auto-generated documentation. Probably for the best, anyways. We'll include more specific documentation and examples later down the line.
- Implements `RpcMethod` for `&T` so you can `call(request)` and optionally, `call(&request)` depending on the context you're in
- Adds a raw error parser for `RpcTransactionError` since it's `error.data` field contains a `ServerError::TxExecutionError`

### CODE REVIEW

**Comment by miraclx** [comment]
Some clear wins:

https://github.com/near/nearcore/commit/e9377113e1150b81a23d19c4c8764ef4e13a4d0d#diff-a69d8e478e63135c7163751e28c8791a504dcea06ae00549752246a5ef8497ecR395-R402

Here, we drop string matching for strictly typed matching 🎉

File: src/methods.rs [hunk 1]
```diff
@@ -1,182 +1,143 @@
 use std::io;
 
+use serde::Deserialize;
 use serde_json::json;
 
 mod chk {
     // this lets us make the RpcMethod trait public but non-implementable by users outside this crate
```
Comment (frol) [inline src/methods.rs:7]: BTW, Why do we believe that is necessary? :thinking:

**Review by frol — APPROVED** [review]
I like more explicit `pub mod ...` since previously I was tricked by the macro syntax

> Here, we drop JSON matching for strictly typed matching

Well, I might argue that we don't want to use any client there at all, and instead construct requests and match results as JSON values to ensure backward compatibility... So far, I don't have a strict preference, but we will see if a strictly typed client actually steps into our way when we extend the API on the server side.

File: src/methods.rs (hunk 1 above)
Comment (miraclx) [reply src/methods.rs:7]: It's used repeatedly inside diff method modules for cases where unit types / null is returned from the server. I didn't want to localize it at the time to avoid littering the file with `use` statements. But I'm currently refactoring so we might not have to anymore.

Comment (frol) [reply src/methods.rs:7]: Hmm, I feel I don't understand your answer or you did not understand my question :smile: 

I meant, why do we need to make RpcMethod non-implementable by users outside of this crate?

Comment (miraclx) [reply src/methods.rs:6]: Oh yeah, sorry.. I was looking at the `serde::Deserialize` highlight 😅 

Well, for one, we have a finite space possible `RpcMethod` types, so I thought if they were completely exhausted in definition within this crate, we don't need to make them implementable outside.

**Comment by miraclx** [comment]
> I like more explicit `pub mod ...` since previously I was tricked by the macro syntax

😅 same, didn't know how much I wanted it until I saw the final syntax

> but we will see if a strictly typed client actually steps into our way when we extend the API on the server side.

Fingers crossed 🤞🏽
